### ISSUE TITLE
Bug: no-unnecessary-condition: false positive with generic union constraint + disable comment not working

### ISSUE THREAD

**Issue by chenglou** [oxc-project/tsgolint#699]
tsgolint 0.14.1, oxlint 1.48.0                                                                                                                                     
                                                                                                                                                                    
**Bug 1**: false positive comparing `T | undefined` to `T`                                                                                                         
                                                                                                                                                                    
```ts
function f<T extends 'a' | 'b'>(x: T | undefined, y: T) {
  return x === y // false positive: "types have no overlap"
}
```

tsgolint resolves T to its constraint "a" | "b" and reports no overlap with T | undefined. eslint's no-unnecessary-condition correctly allows this.

**Bug 2**: eslint-disable-next-line doesn't suppress no-unnecessary-condition

```ts
function g(x: number) {
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (x != null) return x
  return 0
}
```

The disable comment is ignored. Other tsgolint rules like no-unsafe-return respect it fine.

**Comment by chenglou** [comment]
Actually it's 2 bugs in 1. The disabling comment doesn't work either for this rule in particular?

**Comment by camc314** [comment]
> Actually it's 2 bugs in 1. The disabling comment doesn't work either for this rule in particular?

can you share the actual diagnosisic and what it looks like?

**Comment by chenglou** [comment]
```sh
❯ cat src/test-no-unnecessary-condition.ts
// tsgolint 0.14.1, oxlint 1.48.0

// Bug 1: false positive. eslint's no-unnecessary-condition doesn't flag this.
function f<T extends 'a' | 'b'>(x: T | undefined, y: T) {
  return x === y
}

// Bug 2: eslint-disable-next-line doesn't suppress no-unnecessary-condition.
// It works for other tsgolint rules like no-unsafe-return.
function g(x: number) {
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (x != null) return x
  return 0
}
```

```sh
❯ bun oxlint --type-aware src/test-no-unnecessary-condition.ts

  × typescript-eslint(no-unnecessary-condition): This condition will always return the same value since the types have no overlap.
   ╭─[src/test-no-unnecessary-condition.ts:5:9]
 4 │ function f<T extends 'a' | 'b'>(x: T | undefined, y: T) {
 5 │   return x === y
   ·         ─┬    ─┬
   ·          │     ╰── Type: "a" | "b"
   ·          ╰── Type: T | undefined
 6 │ }
   ╰────

  × typescript-eslint(no-unnecessary-condition): This condition will always return the same value since the types have no overlap.
    ╭─[src/test-no-unnecessary-condition.ts:12:7]
 11 │   // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
 12 │   if (x != null) return x
    ·       ┬   ──┬──
    ·       │     ╰── Type: null
    ·       ╰── Type: number
 13 │   return 0
    ╰────

Found 0 warnings and 2 errors.
Finished in 352ms on 1 file with 95 rules using 12 threads.
error: "oxlint" exited with code 1
```

**Comment by camchenry** [comment]
@camc314 I wonder if we need the primary range for the diagnostic for disabling to work correctly? I think this one might only have labeled ranges. We should add one for the `==` or `!=` itself I think and use that as the primary range

**Comment by camc314** [comment]
> @camc314 I wonder if we need the primary range for the diagnostic for disabling to work correctly? I think this one might only have labeled ranges. We should add one for the `==` or `!=` itself I think and use that as the primary range

good idea.

I think we should assume on the oxc side that the first label in the array is the primary one.

**Comment by camchenry** [comment]
> > [@camc314](https://github.com/camc314) I wonder if we need the primary range for the diagnostic for disabling to work correctly? I think this one might only have labeled ranges. We should add one for the `==` or `!=` itself I think and use that as the primary range
> 
> good idea.
> 
> I think we should assume on the oxc side that the first label in the array is the primary one.

I think that's reasonable. We also appear to need to trim these labeled ranges as well as they're including the space before the identifier

**Comment by camc314** [comment]
yep i filed a separate issue for that.

the tsgolint node spans are annoying because they include the leading trivia iirc

**Comment by camc314** [comment]
fixed via #700
